simln-lib/src/lib.rs

```diff
@@ -1085,47 +1109,116 @@ impl<C: Clock + 'static> Simulation<C> {
+        let payment_event_payloads: Arc<Mutex<HashMap<PublicKey, PaymentEventPayload>>> =
+            Arc::new(Mutex::new(HashMap::new()));
         for executor in executors {
-            let sender = producer_channels.get(&executor.source_info.pubkey).ok_or(
-                SimulationError::RandomActivityError(RandomActivityError::ValueError(format!(
-                    "Activity producer for: {} not found.",
-                    executor.source_info.pubkey,
-                ))),
-            )?;
-
-            // pe: produce events
-            let pe_shutdown = self.shutdown_trigger.clone();
-            let pe_listener = self.shutdown_listener.clone();
-            let pe_sender = sender.clone();
-            let clock = self.clock.clone();
+            let payload = PaymentEventPayload {
+                wait_time: Duration::from_secs(0),
+                destination: executor.source_info.clone(),
+                capacity: None,
+                current_count: 0,
+            };
+            payment_event_payloads
+                .lock()
+                .await
+                .insert(executor.source_info.pubkey, payload);
+            generate_payment(&mut heap, executor, 0, payment_event_payloads.clone()).await?;
+        }
 
-            tasks.spawn(async move {
-                let source = executor.source_info.clone();
+        let listener = self.shutdown_listener.clone();
+        let shutdown = self.shutdown_trigger.clone();
+        let clock = self.clock.clone();
+        let task_clone = tasks.clone();
 
-                log::info!(
-                    "Starting activity producer for {}: {}.",
-                    source,
-                    executor.payment_generator
-                );
+        tasks.spawn(async move {
+        loop {
+            select! {
+                biased;
+                _ = listener.clone()  => {
+                    return Ok(())
+                },
+                _ = async {} => {
+                    match heap.pop() {
+                        Some(Reverse(PaymentEvent {
+                            absolute_time: _,
+                            executor,
+                        })) => {
+                            let payload = payment_event_payloads
+                                .lock()
+                                .await
+                                .get(&executor.source_info.pubkey)
+                                .ok_or(SimulationError::PaymentGenerationError(PaymentGenerationError(format!("executor {} not found", executor.source_info))))?
+                                .clone();
+                            if let Some(c) = executor.payment_generator.payment_count() {
+                                if c == payload.current_count {
+                                    log::info!(
+                                        "Payment count has been met for {}: {c} payments. Stopping the activity."
+                                    , executor.source_info);
```

Comment:
nit: just to be aware of, `cargo fmt` breaks in selects for some ungodly reason, so it's worth trying to do a little manual formatting